File: verenigingen/verenigingen/report/expiring_memberships/expiring_memberships.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt


import frappe
from frappe import _
from frappe.utils import getdate


def execute(filters=None):
    columns = get_columns(filters)
    data = get_data(filters)
    return columns, data


def get_columns(filters):
    return [
        _("Membership Type") + ":Link/Membership Type:100",
        _("Membership ID") + ":Link/Membership:140",
        _("Member ID") + ":Link/Member:140",
        _("Member Name") + ":Data:140",
        _("Email") + ":Data:140",
        _("Expiring On") + ":Date:120",
        _("Grace Period") + ":Data:120",
        _("Grace Period Expiry") + ":Date:120",
    ]


def get_data(filters):
    month_name = filters.get("month", "Jan")
    month_number = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"].index(
        month_name
    ) + 1

    # Extract year from fiscal year if it's in format like "2024-2025"
    fiscal_year = filters.get("fiscal_year", "")
    if "-" in fiscal_year:
        year = int(fiscal_year.split("-")[0])
    else:
        try:
            year = int(fiscal_year)
        except (ValueError, TypeError):
            year = getdate().year

    return frappe.db.sql(
        """
        select ms.membership_type, ms.name, m.name, m.full_name, m.email, ms.expiry_date,
               ms.grace_period_status, ms.grace_period_expiry_date
        from `tabMember` m
        inner join (
            select
                memb.name,
                memb.membership_type,
                memb.member,
                memb.grace_period_status,
                memb.grace_period_expiry_date,
                COALESCE(
                    (SELECT next_invoice_date
                     FROM `tabMembership Dues Schedule`
                     WHERE member = memb.member
                     ORDER BY creation DESC
                     LIMIT 1),
                    memb.renewal_date
                ) as expiry_date
            from `tabMembership` memb
            where memb.status in ('Active', 'Pending')
              and (
                  (SELECT next_invoice_date
                   FROM `tabMembership Dues Schedule`
                   WHERE member = memb.member
                   ORDER BY creation DESC
                   LIMIT 1) is not null
                  or memb.renewal_date is not null
              )
        ) ms on m.name = ms.member
        where month(ms.expiry_date) = %(month)s and year(ms.expiry_date) = %(year)s
        order by ms.expiry_date asc
        """,
        {"month": month_number, "year": year},
        as_dict=1,
    )

File: verenigingen/verenigingen/report/expiring_memberships/expiring_memberships.js
/**
 * @fileoverview Expiring Memberships report JavaScript for proactive membership management
 *
 * Provides interactive functionality for the Expiring Memberships report, enabling
 * administrators to proactively manage membership renewals and prevent involuntary
 * membership lapses. This module transforms membership expiration monitoring into
 * an actionable administrative workflow.
 *
 * Key Features:
 * - Visual expiration timeline with urgency indicators
 * - Bulk renewal initiation for multiple memberships
 * - Automated renewal reminder dispatch
 * - Membership extension tools for special cases
 * - Export capabilities for offline processing
 * - Integration with payment processing workflows
 *
 * Renewal Management:
 * - Automated renewal notice generation
 * - Payment link creation for online renewals
 * - Grace period extension for administrative exceptions
 * - Bulk processing for membership batches
 * - Custom renewal terms for special circumstances
 * - Integration with SEPA direct debit workflows
 *
 * Business Context:
 * Essential for maintaining continuous membership engagement and preventing
 * unintentional membership lapses. Helps administrators stay ahead of expiration
 * deadlines, maintain steady membership revenue, and ensure members receive
 * appropriate renewal notices and payment opportunities.
 *
 * Integration:
 * - Connects to Membership DocType for renewal processing
 * - Links to payment processing and invoice generation
 * - Works with email notification systems
 * - Integrates with membership analytics and reporting
 * - Supports member retention tracking and analysis
 *
 * @author Verenigingen Development Team
 * @version 1.4.0
 * @since 2024-04-30
 */

// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Expiring Memberships"] = {
	"filters": [
		{
			"fieldname": "fiscal_year",
			"label": __("Fiscal Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year",
			"default": frappe.defaults.get_user_default("fiscal_year"),
			"reqd": 1
		},
		{
			"fieldname":"month",
			"label": __("Month"),
			"fieldtype": "Select",
			"options": "Jan\nFeb\nMar\nApr\nMay\nJun\nJul\nAug\nSep\nOct\nNov\nDec",
			"default": ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov",
				"Dec"][frappe.datetime.str_to_obj(frappe.datetime.get_today()).getMonth()],
		}
	]
}
